File: tests/Patches/ContributionTest.php
<?php

declare(strict_types=1);

namespace Upkeep\Tests\Patches;

use PHPUnit\Framework\Attributes\DataProvider;
use PHPUnit\Framework\TestCase;
use Upkeep\Drupal\Issue;
use Upkeep\Drupal\IssueFile;
use Upkeep\Drupal\IssueStatus;
use Upkeep\Gitlab\MergeRequest;
use Upkeep\Patches\Contribution;
use Upkeep\Patches\ContributionKind;

final class ContributionTest extends TestCase
{
    /** @param list<IssueFile> $files */
    private static function issue(array $files = [], int $nid = 3489012): Issue
    {
        return new Issue(
            nid: $nid,
            title: 'Add config schema for settings form',
            status: IssueStatus::NeedsReview,
            url: 'https://www.drupal.org/node/3489012',
            project: 'widget',
            priority: 200,
            version: '2.0.x-dev',
            component: 'Code',
            category: 'Task',
            files: $files,
        );
    }

    private static function patch(): IssueFile
    {
        return new IssueFile('3489012-12-config-schema.patch', 'https://example.test/p.patch', 3072, 1705400000);
    }

    /**
     * $base === $head is an empty MR; differing shas carry changes; null
     * leaves it unknown (a list payload, or a snapshot from before diff refs
     * were cached).
     */
    private static function mr(
        int $iid,
        ?string $base = null,
        ?string $head = null,
        string $state = 'opened',
        ?string $mergedAt = null,
        ?string $updatedAt = null,
        string $title = 'Issue #3489012: do the thing',
    ): MergeRequest {
        return new MergeRequest(
            iid: $iid,
            title: $title,
            state: $state,
            authorUsername: 'alice',
            authorId: 100,
            sourceBranch: '3489012-do-the-thing',
            targetBranch: '2.0.x',
            draft: false,
            detailedMergeStatus: null,
            headSha: $head,
            webUrl: 'https://git.drupalcode.org/project/widget/-/merge_requests/' . $iid,
            updatedAt: $updatedAt,
            diffBaseSha: $base,
            diffHeadSha: $head,
            mergedAt: $mergedAt,
        );
    }

    private static function substantiveMr(int $iid = 7): MergeRequest
    {
        return self::mr($iid, 'base111', 'head222');
    }

    private static function emptyMr(int $iid = 1): MergeRequest
    {
        return self::mr($iid, 'same333', 'same333');
    }

    private static function unknownMr(int $iid = 9): MergeRequest
    {
        return self::mr($iid, null, null);
    }

    /** @return iterable<string, array{list<IssueFile>, list<MergeRequest>, ContributionKind}> */
    public static function classificationCases(): iterable
    {
        yield 'a patch and nothing else' => [
            [self::patch()], [], ContributionKind::PatchOnly,
        ];

        yield 'a patch beside a branch that carries changes' => [
            [self::patch()], [self::substantiveMr()], ContributionKind::PatchAndMergeRequest,
        ];

        yield 'a patch beside an empty merge request' => [
            [self::patch()], [self::emptyMr()], ContributionKind::PatchWithEmptyMergeRequest,
        ];

        yield 'a patch beside several merge requests, all empty' => [
            [self::patch()],
            [self::emptyMr(1), self::emptyMr(2)],
            ContributionKind::PatchWithEmptyMergeRequest,
        ];

        yield 'a patch beside one empty and one real merge request' => [
            [self::patch()],
            [self::emptyMr(1), self::substantiveMr()],
            ContributionKind::PatchAndMergeRequest,
        ];

        yield 'no patch, work on a branch' => [
            [], [self::substantiveMr()], ContributionKind::MergeRequestOnly,
        ];

        yield 'no patch, and the only merge request is empty' => [
            [], [self::emptyMr()], ContributionKind::Nothing,
        ];

        yield 'no patch and no merge request' => [
            [], [], ContributionKind::Nothing,
        ];

        // Unknown emptiness reads as real work: a list payload omits diff refs
        // and so does a snapshot written before they were cached, and neither
        // absence is evidence that a branch is empty.
        yield 'a patch beside a merge request of unknown emptiness' => [
            [self::patch()], [self::unknownMr()], ContributionKind::PatchAndMergeRequest,
        ];

        yield 'no patch, and a merge request of unknown emptiness' => [
            [], [self::unknownMr()], ContributionKind::MergeRequestOnly,
        ];

        // A non-patch attachment is not a contribution: an interdiff, a
        // screenshot or a test-only file leaves the issue with nothing to
        // apply.
        yield 'a screenshot is not a patch' => [
            [new IssueFile('before.png', 'https://example.test/before.png', 900, 1705400000)],
            [],
            ContributionKind::Nothing,
        ];
    }

    /**
     * @param list<IssueFile>    $files
     * @param list<MergeRequest> $mrs
     */
    #[DataProvider('classificationCases')]
    public function testClassifiesHowTheWorkArrived(array $files, array $mrs, ContributionKind $expected): void
    {
        $contribution = new Contribution('widget', self::issue($files), $mrs);

        self::assertSame($expected, $contribution->kind());
    }

    /**
     * Only one kind is the dashboard's business rather than this command's.
     */
    public function testOnlyMergeRequestOnlyWorkIsCoveredByTheDashboard(): void
    {
        foreach (ContributionKind::cases() as $kind) {
            self::assertSame(
                $kind === ContributionKind::MergeRequestOnly,
                $kind->isCoveredByMergeRequest(),
                $kind->name,
            );
        }
    }

    public function testKindsCarryingARealBranchAreTheTwoWithOne(): void
    {
        self::assertTrue(ContributionKind::PatchAndMergeRequest->hasSubstantiveMergeRequest());
        self::assertTrue(ContributionKind::MergeRequestOnly->hasSubstantiveMergeRequest());
        self::assertFalse(ContributionKind::PatchOnly->hasSubstantiveMergeRequest());
        self::assertFalse(ContributionKind::PatchWithEmptyMergeRequest->hasSubstantiveMergeRequest());
        self::assertFalse(ContributionKind::Nothing->hasSubstantiveMergeRequest());
    }

    /**
     * The withheld kind is the one kind that never reaches the summary line,
     * because it never reaches the table either.
     */
    public function testEveryKindTheTableCanShowHasASummaryLabel(): void
    {
        foreach (ContributionKind::cases() as $kind) {
            self::assertSame(
                !$kind->isCoveredByMergeRequest(),
                $kind->summaryLabel() !== null,
                $kind->name,
            );
        }
    }

    /** @return iterable<string, array{list<MergeRequest>, string}> */
    public static function mergeRequestCellCases(): iterable
    {
        yield 'no merge request' => [[], '–'];
        yield 'one that carries changes' => [[self::substantiveMr()], '!7'];
        yield 'one that carries nothing' => [[self::emptyMr()], '!1 empty'];
        yield 'one of unknown emptiness is not flagged' => [[self::unknownMr()], '!9'];

        // The substantive one represents the row wherever it is in the list,
        // so a real branch is never reported as empty because a bot draft was
        // listed first.
        yield 'an empty one listed before a real one' => [
            [self::emptyMr(1), self::substantiveMr()], '!7 +1',
        ];
        yield 'a real one listed before an empty one' => [
            [self::substantiveMr(), self::emptyMr(1)], '!7 +1',
        ];
        yield 'two empty ones' => [[self::emptyMr(1), self::emptyMr(2)], '!1 empty +1'];
    }

    /**
     * @param list<MergeRequest> $mrs
     */
    #[DataProvider('mergeRequestCellCases')]
    public function testTheMergeRequestCellNamesTheRepresentativeBranch(array $mrs, string $expected): void
    {
        $contribution = new Contribution('widget', self::issue([self::patch()]), $mrs);

        self::assertSame($expected, $contribution->mergeRequestCell());
    }

    public function testSubstantiveMergeRequestsDropsOnlyTheProvablyEmptyOnes(): void
    {
        $contribution = new Contribution('widget', self::issue(), [
            self::emptyMr(1),
            self::substantiveMr(7),
            self::unknownMr(9),
        ]);

        self::assertSame(
            [7, 9],
            array_map(
                static fn (MergeRequest $mr): int => $mr->iid,
                $contribution->substantiveMergeRequests(),
            ),
        );
    }

    // ------------------------------------------------ has this already landed

    /**
     * The question a Project Update Bot compatibility issue cannot answer
     * itself. Convention keeps those open so the bot can post again as core
     * moves, so an open one may have had its work merged months ago — and two
     * such issues look identical until you ask whether anything was merged.
     *
     * Verified against live data: conditions_helper #3596502 is "active" with
     * its MR merged 2026-06-12, while field_visibility_conditions #3598272 is
     * "needs review" with an open draft. Opposite states, same appearance.
     */
    public function testAMergedMergeRequestIsReportedAsLanded(): void
    {
        $contribution = new Contribution('widget', self::issue([]), [
            self::mr(1, state: 'merged', mergedAt: '2026-06-12T19:48:54.079Z'),
        ]);

        self::assertSame('!1 merged 2026-06-12', $contribution->mergeRequestCell());
        self::assertFalse($contribution->hasWorkNewerThanLanding());
    }

    /**
     * A bot that posts again after its work merged has raised new work. This
     * is why the cell never says "resolved": both readings are statements
     * about evidence, and which one warrants closing stays the maintainer's.
     */
    public function testAPatchPostedAfterTheMergeIsNewerWork(): void
    {
        $afterTheMerge = strtotime('2026-07-01T00:00:00Z');

        $contribution = new Contribution(
            'widget',
            self::issue([new IssueFile('reroll.patch', 'https://example.test/reroll.patch', 10, $afterTheMerge)]),
            [self::mr(1, state: 'merged', mergedAt: '2026-06-12T19:48:54.079Z')],
        );

        self::assertTrue($contribution->hasWorkNewerThanLanding());
        self::assertStringContainsString('newer work since', $contribution->mergeRequestCell());
    }

    /** An open MR touched after the merge is newer work too. */
    public function testAnOpenMergeRequestUpdatedAfterTheMergeIsNewerWork(): void
    {
        $contribution = new Contribution('widget', self::issue([]), [
            self::mr(1, state: 'merged', mergedAt: '2026-06-12T19:48:54.079Z'),
            self::mr(2, state: 'opened', updatedAt: '2026-07-02T10:00:00Z'),
        ]);

        self::assertTrue($contribution->hasWorkNewerThanLanding());
    }

    /** The most recent landing is the one reported. */
    public function testTheLatestMergeIsTheOneReported(): void
    {
        $contribution = new Contribution('widget', self::issue([]), [
            self::mr(1, state: 'merged', mergedAt: '2025-01-01T00:00:00Z'),
            self::mr(9, state: 'merged', mergedAt: '2026-06-12T19:48:54.079Z'),
        ]);

        self::assertSame(9, $contribution->landed()?->iid);
    }

    /** Nothing merged is nothing landed, and the cell reads as it always did. */
    public function testAnIssueWithNothingMergedHasNotLanded(): void
    {
        $contribution = new Contribution('widget', self::issue([]), [self::mr(2, state: 'opened')]);

        self::assertNull($contribution->landed());
        self::assertFalse($contribution->hasWorkNewerThanLanding());
        self::assertStringNotContainsString('merged', $contribution->mergeRequestCell());
    }

    /** A merge with no timestamp cannot be compared against, so nothing is newer. */
    public function testAMergeWithNoTimestampClaimsNoNewerWork(): void
    {
        $contribution = new Contribution('widget', self::issue([]), [self::mr(1, state: 'merged')]);

        self::assertFalse($contribution->hasWorkNewerThanLanding());
    }

    public function testAnUnparseableMergeTimestampClaimsNoNewerWork(): void
    {
        $contribution = new Contribution('widget', self::issue([]), [
            self::mr(1, state: 'merged', mergedAt: 'not a date at all'),
        ]);

        self::assertFalse($contribution->hasWorkNewerThanLanding());
    }

    // ------------------------------------------------- pairing via the fork

    /**
     * The only thing that pairs a Project Update Bot merge request to its
     * issue. Live: the MR is titled "Automated Project Update Bot fixes" on a
     * branch called project-update-bot-only, and its description says only
     * "Relates to #NNN" — which extractOwning() rejects on purpose so the bot
     * cannot suppress an issue's patches by mentioning it. Correct, and it
     * left every bot MR paired to nothing.
     *
     * The fork path is a fact about how the repository came to exist:
     * drupal.org made issue/<module>-<nid> *for* that issue.
     */
    public function testABotMergeRequestPairsThroughItsIssueFork(): void
    {
        $bot = new MergeRequest(
            iid: 1,
            title: 'Automated Project Update Bot fixes',
            state: 'merged',
            authorUsername: 'project update bot',
            authorId: 3644742,
            sourceBranch: 'project-update-bot-only',
            targetBranch: '1.0.x',
            draft: false,
            detailedMergeStatus: null,
            headSha: null,
            webUrl: 'https://git.drupalcode.org/project/widget/-/merge_requests/1',
            description: 'Relates to #3596502. This merge request was automatically created by the bot.',
            sourceProjectId: 216803,
            mergedAt: '2026-06-12T19:48:54.079Z',
        );
        $issue = self::issue([], nid: 3596502);

        // Without the fork map it pairs to nothing at all.
        self::assertSame([], Contribution::pair('widget', [$issue], [$bot])[0]->mergeRequests);

        $paired = Contribution::pair('widget', [$issue], [$bot], [216803 => 3596502]);

        self::assertCount(1, $paired[0]->mergeRequests);
        self::assertSame('!1 merged 2026-06-12', $paired[0]->mergeRequestCell());
    }

    /** A fork map that says nothing about an MR leaves the old rules in charge. */
    public function testTheOldRulesStillApplyWhereTheForkMapIsSilent(): void
    {
        $mr = self::mr(7, title: 'Issue #3467675: Make URL field required');

        $paired = Contribution::pair('widget', [self::issue([], nid: 3467675)], [$mr], [999 => 111111]);

        self::assertCount(1, $paired[0]->mergeRequests);
    }
}

File: go/internal/patches/contribution_test.go
package patches

import (
	"strings"
	"testing"

	"github.com/owenbush/upkeep/internal/drupal"
	"github.com/owenbush/upkeep/internal/gitlab"
)

func realMr(iid int, title, branch string) gitlab.MergeRequest {
	return gitlab.MergeRequest{
		IID: iid, State: "opened", Title: title, SourceBranch: branch,
		DiffBaseSHA: "base", DiffHeadSHA: "head",
	}
}

func emptyMr(iid int, title, branch string) gitlab.MergeRequest {
	mr := realMr(iid, title, branch)
	mr.DiffBaseSHA, mr.DiffHeadSHA = "same", "same"

	return mr
}

// The fork is a fact about how the repository exists, rather than a string in
// a title — and it is the only thing that pairs a Project Update Bot merge
// request at all.
func TestTheForkOutranksEverySignalInTheMetadata(t *testing.T) {
	bot := realMr(1, "Automated Project Update Bot fixes", "project-update-bot-only")
	bot.Description = "Relates to #9999999"
	bot.SourceProjectID = 501

	issue := drupal.Issue{Nid: 3603341, Title: "Drupal 12 compatibility"}
	paired := Pair("pathauto", []drupal.Issue{issue}, []gitlab.MergeRequest{bot}, map[int]int{501: 3603341})

	if len(paired) != 1 || len(paired[0].MergeRequests) != 1 {
		t.Fatalf("got %+v — the bot merge request paired to nothing", paired)
	}
}

// "Relates to #NNN" is rejected by design, so a bot cannot suppress an issue's
// patches by mentioning it.
func TestAMereMentionDoesNotClaimAnIssue(t *testing.T) {
	mentioning := realMr(1, "Something else entirely", "some-branch")
	mentioning.Description = "Relates to #3603341"

	issue := drupal.Issue{Nid: 3603341}
	paired := Pair("pathauto", []drupal.Issue{issue}, []gitlab.MergeRequest{mentioning}, nil)

	if len(paired[0].MergeRequests) != 0 {
		t.Errorf("a mention claimed the issue: %+v", paired[0].MergeRequests)
	}
}

func TestATitleClaimingAnIssuePairsWithoutAFork(t *testing.T) {
	claiming := realMr(1, "Issue #3603341 by owenbush: Drupal 12 compatibility", "3603341-fix")

	issue := drupal.Issue{Nid: 3603341}
	paired := Pair("pathauto", []drupal.Issue{issue}, []gitlab.MergeRequest{claiming}, nil)

	if len(paired[0].MergeRequests) != 1 {
		t.Errorf("got %+v", paired[0].MergeRequests)
	}
}

// Only the issues asked about get a contribution, so a merge request claiming
// one outside the set cannot reach any row.
func TestMergeRequestsClaimingAnIssueOutsideTheSetAreDropped(t *testing.T) {
	elsewhere := realMr(1, "Issue #9999999 by somebody: Another thing", "9999999-fix")

	issue := drupal.Issue{Nid: 3603341}
	paired := Pair("pathauto", []drupal.Issue{issue}, []gitlab.MergeRequest{elsewhere}, nil)

	if len(paired) != 1 {
		t.Fatalf("got %d contributions", len(paired))
	}
	if len(paired[0].MergeRequests) != 0 {
		t.Errorf("got %+v", paired[0].MergeRequests)
	}
}

// Every issue gets a contribution, whether or not anything claims it.
func TestEveryIssueGetsAContribution(t *testing.T) {
	issues := []drupal.Issue{{Nid: 1}, {Nid: 2}, {Nid: 3}}

	paired := Pair("pathauto", issues, nil, nil)
	if len(paired) != 3 {
		t.Fatalf("got %d contributions for %d issues", len(paired), len(issues))
	}
	for _, contribution := range paired {
		if contribution.Module != "pathauto" {
			t.Errorf("module %q", contribution.Module)
		}
	}
}

// Unknown emptiness counts as real work: that is the reading a list payload
// gives, and inventing empty merge requests out of missing data is the wrong
// direction.
func TestUnknownEmptinessCountsAsSubstantive(t *testing.T) {
	unknown := realMr(1, "t", "b")
	unknown.DiffBaseSHA, unknown.DiffHeadSHA = "", ""

	substantive := Substantive([]gitlab.MergeRequest{unknown, emptyMr(2, "t", "b")})
	if len(substantive) != 1 || substantive[0].IID != 1 {
		t.Errorf("got %+v", substantive)
	}
}

func TestTheKindFollowsWhatActuallyArrived(t *testing.T) {
	withPatch := drupal.Issue{Nid: 1, Files: []drupal.IssueFile{patch("a.patch", 100, 10)}}
	withoutPatch := drupal.Issue{Nid: 1}

	for _, tc := range []struct {
		name          string
		issue         drupal.Issue
		mergeRequests []gitlab.MergeRequest
		want          Kind
	}{
		{"patch alone", withPatch, nil, PatchOnly},
		{"patch and a real branch", withPatch, []gitlab.MergeRequest{realMr(1, "t", "b")}, PatchAndMergeRequest},
		{"patch and an empty draft", withPatch, []gitlab.MergeRequest{emptyMr(1, "t", "b")}, PatchWithEmptyMergeRequest},
		{"branch alone", withoutPatch, []gitlab.MergeRequest{realMr(1, "t", "b")}, MergeRequestOnly},
		{"nothing at all", withoutPatch, nil, Nothing},
		{"an empty draft and nothing else", withoutPatch, []gitlab.MergeRequest{emptyMr(1, "t", "b")}, Nothing},
	} {
		contribution := Contribution{Module: "pathauto", Issue: tc.issue, MergeRequests: tc.mergeRequests}
		if got := contribution.Kind(); got != tc.want {
			t.Errorf("%s: got %v, want %v", tc.name, got, tc.want)
		}
	}
}

// Only the MR-only kind is reachable from the dashboard, and so has no
// business on a patch report.
func TestOnlyMergeRequestOnlyIsCoveredElsewhere(t *testing.T) {
	if !MergeRequestOnly.IsCoveredByMergeRequest() {
		t.Error("an MR-only issue is not covered by the dashboard")
	}
	for _, kind := range []Kind{PatchOnly, PatchAndMergeRequest, PatchWithEmptyMergeRequest, Nothing} {
		if kind.IsCoveredByMergeRequest() {
			t.Errorf("%v was treated as covered", kind)
		}
		if kind.SummaryLabel() == "" {
			t.Errorf("%v has no summary label", kind)
		}
	}
	if MergeRequestOnly.SummaryLabel() != "" {
		t.Error("an MR-only issue is summarised, and it should not be")
	}
}

// A landing outranks everything else the cell could say: it is the one thing a
// maintainer cannot read off the issue at all.
func TestALandingOutranksEverythingElseInTheCell(t *testing.T) {
	landed := realMr(7, "t", "b")
	landed.State = "merged"
	landed.MergedAt = "2026-06-12T10:00:00Z"

	open := realMr(9, "t", "b")

	cell := RenderMergeRequestCell([]gitlab.MergeRequest{open, landed}, &landed, false)
	if !strings.Contains(cell, "!7 merged 2026-06-12") {
		t.Errorf("cell %q", cell)
	}
	if strings.Contains(cell, "!9") {
		t.Errorf("cell %q buries the landing behind the open one", cell)
	}
}

func TestNewerWorkSinceALandingIsSaid(t *testing.T) {
	landed := realMr(7, "t", "b")
	landed.State = "merged"
	landed.MergedAt = "2026-06-12T10:00:00Z"

	cell := RenderMergeRequestCell([]gitlab.MergeRequest{landed}, &landed, true)
	if !strings.Contains(cell, "newer work since") {
		t.Errorf("cell %q", cell)
	}
}

// A bot that posts again after its earlier work merged has raised new work.
func TestAPatchNewerThanTheMergeIsNewerWork(t *testing.T) {
	landed := realMr(7, "t", "b")
	landed.State = "merged"
	landed.MergedAt = "2026-06-12T10:00:00Z"

	// 2026-06-13, the day after.
	after := drupal.Issue{Nid: 1, Files: []drupal.IssueFile{patch("re-roll.patch", 1781308800, 10)}}
	// 2026-01-01, months before.
	before := drupal.Issue{Nid: 1, Files: []drupal.IssueFile{patch("original.patch", 1767225600, 10)}}

	newer := Contribution{Issue: after, MergeRequests: []gitlab.MergeRequest{landed}}
	if !newer.HasWorkNewerThanLanding() {
		t.Error("a patch posted after the merge did not read as newer work")
	}

	older := Contribution{Issue: before, MergeRequests: []gitlab.MergeRequest{landed}}
	if older.HasWorkNewerThanLanding() {
		t.Error("a patch posted before the merge read as newer work")
	}
}

// An open merge request touched after the landing is newer work too; the
// merged one itself is not.
func TestAnOpenMergeRequestTouchedAfterTheLandingCounts(t *testing.T) {
	landed := realMr(7, "t", "b")
	landed.State = "merged"
	landed.MergedAt = "2026-06-12T10:00:00Z"
	// Later than the merge, which is ordinary: a comment or a label after
	// merging bumps updated_at again. Without the state check the merged merge
	// request would report itself as work newer than its own landing.
	landed.UpdatedAt = "2026-07-01T09:00:00Z"

	open := realMr(9, "t", "b")
	open.UpdatedAt = "2026-09-01T10:00:00Z"

	withOpen := Contribution{MergeRequests: []gitlab.MergeRequest{landed, open}}
	if !withOpen.HasWorkNewerThanLanding() {
		t.Error("an open merge request updated after the merge did not count")
	}

	// The merged one alone must not count itself: merging bumps updated_at.
	aloneCase := Contribution{MergeRequests: []gitlab.MergeRequest{landed}}
	if aloneCase.HasWorkNewerThanLanding() {
		t.Error("the merged merge request counted itself as newer work")
	}
}

func TestNothingLandedIsNotNewerWork(t *testing.T) {
	contribution := Contribution{MergeRequests: []gitlab.MergeRequest{realMr(1, "t", "b")}}

	if contribution.Landed() != nil {
		t.Error("an open merge request read as landed")
	}
	if contribution.HasWorkNewerThanLanding() {
		t.Error("newer work than a landing that did not happen")
	}
}

// An issue can carry both a real branch and a bot's empty draft, and the real
// branch is the answer to "is this already in git?".
func TestARealBranchRepresentsTheIssueAheadOfAnEmptyDraft(t *testing.T) {
	cell := RenderMergeRequestCell([]gitlab.MergeRequest{emptyMr(1, "t", "b"), realMr(2, "t", "b")}, nil, false)

	if !strings.HasPrefix(cell, "!2") {
		t.Errorf("cell %q, want the real branch first", cell)
	}
	if strings.Contains(cell, "empty") {
		t.Errorf("cell %q flags the representative as empty when it is not", cell)
	}
	if !strings.Contains(cell, "+1") {
		t.Errorf("cell %q does not count the other", cell)
	}
}

// The bug this closed: a row showing an empty merge request and a patch used
// to suggest a check that would do nothing.
func TestAnEmptyRepresentativeIsFlagged(t *testing.T) {
	cell := RenderMergeRequestCell([]gitlab.MergeRequest{emptyMr(1, "t", "b")}, nil, false)

	if cell != "!1 empty" {
		t.Errorf("cell %q", cell)
	}
}

func TestNoMergeRequestsIsADash(t *testing.T) {
	if got := RenderMergeRequestCell(nil, nil, false); got != "–" {
		t.Errorf("got %q", got)
	}
}

// Nothing here is mergeable, and a cell that looked like a gate verdict would
// invite the wrong action.
func TestTheDashboardStatusSaysWhatArrived(t *testing.T) {
	one := drupal.Issue{Nid: 1, Files: []drupal.IssueFile{patch("a.patch", 100, 10)}}
	two := drupal.Issue{Nid: 1, Files: []drupal.IssueFile{
		patch("a.patch", 100, 10), patch("b.patch", 200, 10),
	}}

	if got := (Contribution{Issue: one}).DashboardStatus(); got != "PATCH 1 patch" {
		t.Errorf("got %q", got)
	}
	if got := (Contribution{Issue: two}).DashboardStatus(); got != "PATCH 2 patches" {
		t.Errorf("got %q", got)
	}
	if got := (Contribution{Issue: drupal.Issue{Nid: 1}}).DashboardStatus(); got != "PATCH nothing attached" {
		t.Errorf("got %q", got)
	}

	covered := Contribution{Issue: drupal.Issue{Nid: 1}, MergeRequests: []gitlab.MergeRequest{realMr(4, "t", "b")}}
	if got := covered.DashboardStatus(); !strings.HasPrefix(got, "PATCH covered by !4") {
		t.Errorf("got %q", got)
	}
}

// Nothing a cached result could be about.
func TestAnIssueWithNoPatchHasNoCurrentRevision(t *testing.T) {
	if got := (Contribution{Issue: drupal.Issue{Nid: 1}}).CurrentRevision(); got != "" {
		t.Errorf("got %q", got)
	}
}

func TestTheCurrentRevisionIsTheNewestPatchs(t *testing.T) {
	issue := drupal.Issue{Nid: 1, Files: []drupal.IssueFile{
		patch("old.patch", 100, 10),
		patch("new.patch", 200, 10),
	}}

	contribution := Contribution{Issue: issue}
	if got := contribution.CurrentRevision(); got != Revision("https://www.drupal.org/files/issues/new.patch") {
		t.Errorf("got %q, want the newest patch's", got)
	}
}
